File: web/1/lib/dto.go
package lib

type DotParams struct {
	X string  `json:"X"`
	Y float64 `json:"Y"`
	R string  `json:"R"`
}

type DotStatus struct {
	Entry    DotParams `json:"entry"`
	Hit      bool      `json:"hit"`
	Date     string    `json:"date"`
	Duration string    `json:"duration"`
}

File: web/1/src/dto.ts
export type DotParams = {
  X: string;
  Y: number;
  R: string;
}

export type DotStatus = {
  entry: DotParams;
  hit: boolean;
};

export function isDotParams(object: unknown): object is DotParams {
  return object instanceof Object &&
    'X' in object && typeof object['X'] === 'string' &&
    'Y' in object && typeof object['Y'] === 'number' &&
    'R' in object && typeof object['R'] === 'string';
}

export function isDotStatus(object: unknown): object is DotStatus {
  return object instanceof Object &&
    'entry' in object && isDotParams(object['entry']);
}